spock-specs/src/test/groovy/org/spockframework/util/ReflectionUtilSpec.groovy

```diff
@@ -200,4 +205,116 @@ class ReflectionUtilSpec extends Specification {
   static class Generics<T> {
     void foo(T one, List<T> two, List<String> three) {}
   }
+
+  def "isAnnotationPresentRecursive"(Class<?> clazz, Class<? extends Annotation> annotation, boolean expectedResult) {
+    expect:
+    ReflectionUtil.isAnnotationPresentRecursive(clazz, annotation) == expectedResult
+    where:
+    clazz              | annotation | expectedResult
+    Object             | Nullable   | false
+    Specification      | Testable   | true
+    ReflectionUtilSpec | Testable   | true
+    ReflectionUtilSpec | Nullable   | false
+    TypeWithAnno       | Issue      | true
+    SubTypeNoAnno      | Issue      | true
+  }
+
+  def "getAnnotationRecursive"(Class<?> clazz, Class<? extends Annotation> annotationClass, boolean expectedResult) {
+    when:
+    def anno = ReflectionUtil.getAnnotationRecursive(clazz, annotationClass)
+    if (expectedResult) {
+      assert clazz && annotationClass && anno != null
+    } else {
+      assert clazz && annotationClass && anno == null
+    }
+    then:
+    noExceptionThrown()
+    where:
+    clazz              | annotationClass | expectedResult
+    Object             | Nullable        | false
+    Specification      | Testable        | true
+    ReflectionUtilSpec | Testable        | true
+    ReflectionUtilSpec | Nullable        | false
+    TypeWithAnno       | Issue           | true
+    SubTypeNoAnno      | Issue           | true
+  }
+
+  def "collectAnnotationRecursive"(Class<?> clazz, Class<? extends Annotation> annotationClass, int expectedCount) {
+    expect:
+    ReflectionUtil.collectAnnotationRecursive(clazz, annotationClass).size() == expectedCount
+    where:
+    clazz              | annotationClass | expectedCount
+    Object             | Nullable        | 0
+    Specification      | Testable        | 1
+    ReflectionUtilSpec | Testable        | 2
+    ReflectionUtilSpec | Nullable        | 0
+    TypeWithAnno       | Issue           | 1
+    SubTypeNoAnno      | Issue           | 1
+    SubTypeWithAnno    | Issue           | 2
+  }
+
+  @Issue("")
```

Comment:
I'd use `Narrative` instead of `Issue`, or another annotation, as  we might do some data mining with `Issue` and this would just add noise.